### PR TITLE
Add Grid interactivity

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Adds a new "Grid interactivity" experiment and adds new functionally behind this feature flag that lets the user:

- Click on a Grid block and see a dotted outline of where items will be placed. (`GridVisualizer`)
- Click on an item in a Grid block and set the column and row span using drag handles. (`GridItemResizer`)

## Why?
https://github.com/WordPress/gutenberg/issues/57478

## How?
- `GridVisualizer`
  - Pretty similar to the existing `MarginVisualizer` and `PaddingVisualizer`.
  - Is rendered when a Grid block or one of its children are selected.
  - Uses `getComputedStyle` to determine the computed `grid-template` and uses this to lay out boxes containing a dotted border.
- `GridItemResizer`
  - Renders a `ResizableBox` on top of the grid item.
  - When the user finishes resizing, we set the span by looking for the index of the nearest _grid line_ (see https://css-tricks.com/snippets/css/complete-guide-grid/ for a refresher on grid terminology) to the cursor. The span is the start grid line subtracted from the end grid line.

## Testing Instructions
1. Enable the Grid interactivity experiment
2. Insert a Grid block
3. Select the Grid block or one of its child blocks
4. Go nuts

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/612155/bccd1754-32d5-4c3c-ae07-07a4b7021dd0

### CODE REVIEW

File: packages/block-editor/src/hooks/layout.scss [hunk 1]
```diff
@@ -39,3 +39,16 @@
...
+.block-editor-grid-visualizer .components-popover__content * {
+	pointer-events: none !important;
+}
+
+.block-editor-grid-visualizer__grid {
+	display: grid;
+}
+
+.block-editor-grid-visualizer__item {
+	border: 1px dashed $gray-300;
```
Comment (hanneslsm) [inline packages/block-editor/src/hooks/layout.scss]: This will not be visible when the theme background is greyish. I'd go with a `color-mix(in srgb, var(--wp-admin-theme-color) 20%, transparent);` or so.

**Comment by tellthemachines** [comment]
This'll need a rebase now the column span refactor is merged. 

I'm playing around with a grid of images and the duplicated handles are a bit confusing:
Screenshot 2024-02-16 at 3 00 38 pm

Would it make sense to hide the Image block resize drag handles when inside a grid?

File: packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	const blockElement = useBlockElement( clientId );
+	if ( ! blockElement ) {
+		return null;
+	}
+	return (
+		<BlockPopover
+			className="block-editor-grid-item-resizer"
+			clientId={ clientId }
+			__unstableCoverTarget
+			__unstablePopoverSlot="block-toolbar"
```
Comment (tellthemachines) [inline packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:22]: Is the use of the toolbar slot the reason why the resize handles move around when scrolling? 
Screenshot 2024-02-16 at 3 18 24 pm

File: packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+					const boxWidth = element.offsetWidth;
+					const boxHeight = element.offsetHeight;
+					const { columnGap, rowGap, itemWidth, itemHeight } =
+						getGridDimensions( blockElement.parentElement );
+					const columnSpan = Math.round(
+						( columnGap + boxWidth ) / ( columnGap + itemWidth )
+					);
+					const rowSpan = Math.round(
+						( rowGap + boxHeight ) / ( rowGap + itemHeight )
+					);
```
Comment (tellthemachines) [inline packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:71]: Hmmm I'm not sure we can use the same logic for columns as we do for rows. The problem is that our grid doesn't define an intrinsic height for its rows (which is fine because we want rows to adapt to whatever the content height is). But that means that pulling a block down to expand over non-existing rows will cause a whole bunch of collapsed rows to appear:
Screenshot 2024-02-16 at 3 42 44 pm

Maybe we can play around with dynamically changing the height of the element as it's resized (though this may not work too well for elements with a fixed aspect ratio, like images) and only expand the amount of rows by 1 further than the existing grid rows (because, row height being variable, the next row can be as big as we want to so we only ever need to add 1 at a time)

**Review by tellthemachines — COMMENTED** [review]
This is a very cool experiment!

One thing we should probably consider is updating the grid visualisation when parent or child values are changed via the sidebar controls. It might not be a huge deal for the children but for the grid itself I think folks might want to tweak column number etc. as they're manipulating the contents:

https://github.com/WordPress/gutenberg/assets/8096000/df3756c1-9f88-4cde-83e3-d4d53b98c4f6

**Comment by noisysocks** [comment]
Great feedback, thanks @tellthemachines!

File: packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js (hunk 2 above)
Comment (noisysocks) [reply packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:22]: Is it better now? I added `shift={ false }` to the popovers which seems to help.

File: packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js (hunk 3 above)
Comment (noisysocks) [reply packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:71]: Ah yeah you're right. Bummer. Will play around with this. `getComputedStyle` returns a value for `grid-template-rows` that includes the dynamic heights so maybe I can change my approach to calculate the span based on that.

Comment (noisysocks) [reply packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:71]: OK I changed the approach and think this works better.

https://github.com/WordPress/gutenberg/assets/612155/25c9a612-ba4d-458d-ace6-e2569787785d

There's still some weirdness when e.g. you change a row span such that the item appears in a different column. The box doesn't update to reflect the new grid template. I think I can fix that by forcing a re-render, will keep plugging away at it.

File: packages/block-editor/src/components/block-popover/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/block-popover/index.js:1]: Probably best to split these changes out to their own PR as `__unstableCoverBlock` is used in a few places and I'm not sure I fully understand the implications of having it use a `ResizeObserver`.

**Comment by noisysocks** [comment]
Still working on this. When ready I'll put the changes behind an experimental flag and mark the PR as ready for review.

File: packages/block-editor/src/components/block-popover/index.js
Comment (noisysocks) [reply packages/block-editor/src/components/block-popover/index.js:1]: Done: https://github.com/WordPress/gutenberg/pull/59228

**Comment by noisysocks** [comment]
I've stacked this onto https://github.com/WordPress/gutenberg/pull/59228 which contains a `BlockPopover` fix we'll need to merge first but otherwise think this is ready. I've put everything behind a new experimental flag. Will do some testing tomorrow as it's end of day now.

**Comment by tellthemachines** [comment]
Just playing around with this and noticed that multiple resize actions on the same block only create one undo level:

https://github.com/WordPress/gutenberg/assets/8096000/7d58e0b0-f412-4e59-abd0-0045a92d9bf4

I'm thinking it might be better for each resize to be separately undoable if possible?

Apart from that it's working very smoothly! Only being able to drag right and bottom can be annoying depending on the position of the block being dragged (if it's the bottom right block we have to sort of move it around until it's draggable? Or use the sidebar controls) but this is a pretty great start for an experimental feature!

File: packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				} }
+				enable={ {
+					bottom: true,
+					bottomLeft: false,
+					bottomRight: false,
+					left: false,
+					right: true,
+					top: false,
+					topLeft: false,
+					topRight: false,
```
Comment (tellthemachines) [inline packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:38]: So what happens if we enable "left" and "top"? I'm guessing it breaks somehow? 😄

File: packages/block-editor/src/components/grid-visualizer/utils.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export function getComputedCSS( element, property ) {
```
Comment (tellthemachines) [inline packages/block-editor/src/components/grid-visualizer/utils.js:1]: This is the third instance of this function in the block-editor package, might be worth consolidating (as a separate task)

**Review by tellthemachines — APPROVED** [review]
LGTM! No blockers to merging this as an experiment.

**Review by andrewserong — COMMENTED** [review]
Very cool, this is really fun to play with, too. Not a blocker for this PR I don't think, but something I noticed while playing around is that if the column span on a block is larger than the number of columns available, the manual column number of the parent Grid block doesn't force the blocks down to that number of columns. It's a little hard to describe, so I'll see if a video works:

At the beginning of the video we're adjusting number of column span, which is working nicely. At the end of the video, I go up to the Grid block to reduce the number of columns, but it won't let us go down to, say, 2 columns:

https://github.com/WordPress/gutenberg/assets/14988353/81dc3f74-2261-4d74-902b-52bdddea98d2

Again, not a blocker for landing this experiment, though!

File: packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js (hunk 4 above)
Comment (noisysocks) [reply packages/block-editor/src/components/grid-visualizer/grid-item-resizer.js:38]: I think there's a bug in `ResizableBox` or something because it just goes crazy 🤪 We can look at adding it later if it feels like we need it.

File: packages/block-editor/src/components/grid-visualizer/utils.js (hunk 5 above)
Comment (noisysocks) [reply packages/block-editor/src/components/grid-visualizer/utils.js:1]: The refactor in https://github.com/WordPress/gutenberg/pull/59227 brings it down to 2! Getting close... 😂

**Comment by noisysocks** [comment]
Thanks everyone! Merging this (behind an experimental flag) and will look into those bugs y'all found next week.

**Comment by noisysocks** [comment]
List of follow up work for clarity:

- [ ] DRY up `getComputedStyle` https://github.com/WordPress/gutenberg/pull/59052#discussion_r1500168573
- [ ] Fix bug with manual column adjustment https://github.com/WordPress/gutenberg/pull/59052#pullrequestreview-1897314373
- [ ] Fix bug with undo levels https://github.com/WordPress/gutenberg/pull/59052#issuecomment-1960657273
- [ ] Adjust grid colour so that it's visible on grey backgrounds https://github.com/WordPress/gutenberg/pull/59052#discussion_r1490735623

**Comment by tellthemachines** [comment]
> the manual column number of the parent Grid block doesn't force the blocks down to that number of columns.

This is just how grids work: if a grid child specifies a number of columns greater than what's defined in `grid-template-columns`, the grid just grows itself a bunch of new columns to accommodate the child. For the auto grids we use container queries to mitigate that; for manual grids it's not clear what the behaviour should be. I'm inclined to just let default grid behaviour happen here, because even if someone accidentally makes an item span more columns than the grid has it's pretty easy to notice and fix it.

**Comment by andrewserong** [comment]
> I'm inclined to just let default grid behaviour happen here, because even if someone accidentally makes an item span more columns than the grid has it's pretty easy to notice and fix it.

The case I had in mind is a bit more subtle: I'm thinking of if someone goes to update a pattern or template that already uses a Grid that has some children that are spanning multiple columns, and then they decide to reduce the Grid down to 2 columns. In this case, they mightn't know that there are some children that have spans already set on them, so it'll look like the control isn't doing anything, and they might not know they have to go in to update those children.

Definitely not a blocker for now, but I wondered if it's maybe something that could be smoothed over when the manual column count is updated? I was very much thinking of that as an idea for polish, though, not anything that needs to be fixed up immediately 🙂

**Comment by tellthemachines** [comment]
> if someone goes to update a pattern or template that already uses a Grid that has some children that are spanning multiple columns, and then they decide to reduce the Grid down to 2 columns

A good way to provide visual feedback in that case would be to always show `GridVisualizer` when the grid block is selected, so changing the column count will have a visual effect. We could even check the computed value of `grid-template-columns` for any cols with 0 width and show a warning/highlight them in colour or something.

I can't think of a good way to actually force the child items to span less columns: we'd have to iterate through all the block's inner blocks in search of blocks with spans greater than the column number, and perform updates on them. It's not solvable with a container query because the total width of the grid is always the same regardless of column count.

Because the scenario is a bit of an edge case, it might be better to wait and see if anyone actually runs into it before preemptively fixing it 😅

**Comment by paaljoachim** [comment]
This looks very cool and will become a very powerful layout tool!
It makes me think of the old school HTML tables. With table cells etc.
When I see the dragging of one cell to turn into 3-4-5 I wonder if a specific cell can be merged so that it from containing 3-4-5 cells to only have 1 bigger cell.

In a sense what is being created here will replace the Table block or in another sense bring in a very nice update to the Table block.

What would be nice is an overview video of the whole experiment. Some of the motivation behind it, what the goals are, if this could also relate to device specific break points. There are so many aspect this touches upon. Having an overview video would very much help and also can be shared giving the opportunity to broaden the feedback.

**Comment by noisysocks** [comment]
Hey @paaljoachim. I turned https://github.com/WordPress/gutenberg/issues/57478 into a an overview issue for this focus though it's probably lacking the exact detail you're after. Maybe @SaxonF could elaborate—he has a way with words—but the gist of it is that we want to empower users to be able to create more advanced layouts and more advanced block themes in a more intuitive way than they can currently using Group, Row, Stack, Columns, etc. Grid tools exist in other proprietary site builders and ultimately the goal is for WordPress to take market share from proprietary CMS tools.

There are some parallels between Grid and the Table block, just as there are parallels between `display: grid` and `<table>`, but fundamentally the Table block is intended for _tabular data_ e.g. the kind you'd find in a spreadsheet and this won't change.
